inc/c_util/thandle_tuple_array.h

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#ifndef THANDLE_TUPLE_ARRAY_H
+#define THANDLE_TUPLE_ARRAY_H
+
+#ifdef __cplusplus
+#include <cinttypes>
+#else
+#include <inttypes.h>
+#endif
+
+#include "macro_utils/macro_utils.h"
+
+#include "c_util/thandle.h"
+
+#include "umock_c/umock_c_prod.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define THANDLE_TUPLE_ARRAY_STRUCT_MEMBER(type, name) THANDLE(type) name;
+
+#define DECLARE_THANDLE_TUPLE_TYPE(name, ...) \
+    typedef struct MU_C2(name, _TAG) \
+    { \
+        MU_FOR_EACH_2(THANDLE_TUPLE_ARRAY_STRUCT_MEMBER, __VA_ARGS__) \
+    } name;
+
+#define THANDLE_TUPLE_ARRAY_TYPE(name) MU_C2A(name, _ARRAY)
+
+#define DECLARE_THANDLE_TUPLE_ARRAY_TYPE(name) \
+    typedef struct MU_C2B(THANDLE_TUPLE_ARRAY_TYPE(name), _TAG) \
+    { \
+        const uint32_t count; \
+        name * tuple_array; \
+    } THANDLE_TUPLE_ARRAY_TYPE(name);
+
+#define THANDLE_TUPLE_ARRAY_CREATE(name) MU_C2A(name, _array_create)
+#define THANDLE_TUPLE_ARRAY_DESTROY(name) MU_C2A(name, _array_destroy)
+
+#define DECLARE_THANDLE_TUPLE_ARRAY(name, ...) \
+    DECLARE_THANDLE_TUPLE_TYPE(name, __VA_ARGS__) \
+    DECLARE_THANDLE_TUPLE_ARRAY_TYPE(name) \
+    MOCKABLE_FUNCTION(, THANDLE_TUPLE_ARRAY_TYPE(name)*, THANDLE_TUPLE_ARRAY_CREATE(name), uint32_t, count); \
+    MOCKABLE_FUNCTION(, void, THANDLE_TUPLE_ARRAY_DESTROY(name), THANDLE_TUPLE_ARRAY_TYPE(name)*, tuple_array);
+
+#define THANDLE_TUPLE_ARRAY_INIT_MEMBER(type, name) THANDLE_INITIALIZE(type)(&result->tuple_array[i].name, NULL);
+
+#define IMPLEMENT_THANDLE_TUPLE_ARRAY_CREATE(name, ...) \
+    IMPLEMENT_MOCKABLE_FUNCTION(, THANDLE_TUPLE_ARRAY_TYPE(name)*, THANDLE_TUPLE_ARRAY_CREATE(name), uint32_t, count) \
+    { \
+        THANDLE_TUPLE_ARRAY_TYPE(name)* result; \
+        \
+        /*Codes_SRS_THANDLE_TUPLE_ARRAY_42_001: [ THANDLE_TUPLE_ARRAY_CREATE(name) shall allocate memory for the array. ]*/ \
+        size_t required_size = sizeof(THANDLE_TUPLE_ARRAY_TYPE(name)); \
+        result = (THANDLE_TUPLE_ARRAY_TYPE(name)*)malloc(required_size); \
+        \
+        if (result == NULL) \
+        { \
+            /*Codes_SRS_THANDLE_TUPLE_ARRAY_42_004: [ If there are any errors then THANDLE_TUPLE_ARRAY_CREATE(name) shall fail and return NULL. ]*/ \
+            LogError("malloc(%zu) failed", required_size); \
+        } \
+        else \
+        { \
+            result->tuple_array = NULL; \
+            bool failed = false; \
+            if (count > 0) \
+            { \
+                /*Codes_SRS_THANDLE_TUPLE_ARRAY_42_002: [ THANDLE_TUPLE_ARRAY_CREATE(name) shall allocate memory for count structs which hold the tuple. ]*/ \
+                required_size = count * sizeof(name); \
+                result->tuple_array = (name *)malloc(required_size); \

```

Comment:
>malloc [](http://example.com/codeflow?start=46&length=6)

Same can be achieved using just 1 malloc and flexible arrays. Would imply having line 37 written as name tuple_array[].

Saves 1 malloc.

Reply:
Yes, doesn't hurt to optimize this